### PR TITLE
Generic refactor and improvements

### PR DESCRIPTION
**Pull request by wladekb**
Naming cleanup and logging improvements (preliminary work to support sending structured logs to ELK).

Includes #26 

/cc @pchojnacki @dianafa @owend @SeanColombo @mixth-sense

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -3,18 +3,19 @@ var config = require("./server_config.js");
...
+	, storage = require('./storage').redisFactory()
+	, models = require('./models/models')
+	, mediawikiClient = require('./mw_client.js').MediawikiClient
+	, loggerModule = require('./logger.js')
+	, tracker = require('./tracker.js')
+	, monitoring = require('./monitoring.js')
+	, logger = loggerModule.logger
+		.withAttribute('server', 'module')
+	, keys = require("./keys_helpers")
+	, Promise = require('promise');
```
Comment (dianafa) [inline server.js]: why so strange formatting?

**Comment by pchojnacki** [comment]
```/home/travis/build/Wikia/chat/logger.js
  185:18  error  Move the invocation into the parens that contain the function  wrap-iife
/home/travis/build/Wikia/chat/server.js
  805:9   error  'userData' is already declared in the upper scope  no-shadow
  990:21  error  ["sessionId"] is better written in dot notation    dot-notation
  991:27  error  ["username"] is better written in dot notation     dot-notation
```

File: server.js (hunk 1 above)
Comment (pchojnacki) [inline server.js]: + looks to me Promise is not used here anywhere. Is it necessary ?

Comment (wladekb) [inline server.js]: * It's been like that when I opened the code. I'll change it.
* Yes,... Promise is not used as I intended to convert a lot of code to use promises but I didn't get to actually doing it. I'll remove that.

**Comment by wladekb** [comment]
@pchojnacki tests are now failing even for node v0.12

File: lib/mw.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+module.exports.urlencode = function(str) {
```
Comment (wladekb) [inline lib/mw.js]: it would make more sense to name this file "urlencode.js" instead of "mw.js"

Comment (mixth-sense) [inline lib/mw.js]: @pchojnacki - please, address this remark unless you consider it pointless :-)

Comment (pchojnacki) [inline lib/mw.js]: I've renamed it as I myself had mixed feeling about it name. I renamed it to mw_urlencode as my understanding based on comments is that its somewhat mw specific.

File: controllers/commandLineControllers.js [hunk 3]
```diff
@@ -24,36 +23,36 @@ var NodeChatController = {
...
-            if(NodeChatController.announceConnection){
+
+		this.socket.on('connect', function() {
+			if (NodeChatController.announceConnection) {
 				inlineAlert(NodeChatController.model, 'Reconnected!'); // TODO: i18n: how do we do i18n for this?
 				NodeChatController.announceConnection = false;
 			}
-            NodeChatController.connected = true;
+			NodeChatController.connected = true;
 //            //view.setConnected(true);
```
Comment (hakubo) [inline controllers/commandLineControllers.js]: can we remove commented code?

File: lib/mw_urlencode.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+module.exports.urlencode = function(str) {
+    // http://kevin.vanzonneveld.net
+    // +   original by: Philip Peterson
```
Comment (hakubo) [inline lib/mw_urlencode.js]: is this comment needed here?

File: controllers/commandLineControllers.js [hunk 5]
```diff
@@ -1,10 +1,9 @@
 var WIKIA_NODE_HOST = "http://chatserver.wikia-dev.com";
 var WIKIA_NODE_PORT = 8000;
-io = require('../socket.io/socket.io.js');
-//var io = require('socket.io');
-Backbone = require('backbone');
-var models = require('../models/models.js');
-//var views = require('../views/views.js');
+
+var io = require('../socket.io/socket.io.js'),
```
Comment (hakubo) [inline controllers/commandLineControllers.js]: one var per line?

File: monitoring.js [hunk 6]
```diff
@@ -1,80 +1,84 @@
-var config = require("./server_config.js"); // our node-chat config
-var http = require("http");
-var urlUtil = require('url');
-var logger = require('./logger').logger;
+var config = require("./server_config.js"),
```
Comment (hakubo) [inline monitoring.js]: one var per line?

File: monitoring.js [hunk 7]
```diff
@@ -1,80 +1,84 @@
...
-        // Holy log-spam batman! :)
-        //logger.debug(stats);
-        if(data.indexOf('allocation_stats')) {
-            callback(stats);
-        }    	
-    },
-    errorCallback);
+	var stats = [];
+	storage.info(function(storagedata) {
+			var data = storagedata.split("\r\n");
```
Comment (hakubo) [inline monitoring.js]: one tab too many?

File: monitoring.js [hunk 8]
```diff
@@ -1,80 +1,84 @@
...
-				var CMDstring = 'gmetric --name="' + i + '_' + config.INSTANCE + '" --value='+ out[i] +' --type=float --unit=n --dmax=90 --group=Chat';
-				execCMD(CMDstring, function() {});
+exports.startMonitoring = function(interval, storage) {
+	logger.debug('startMonitoring...');
+	setInterval(function() {
+		getStats(storage, function(out) {
+			for (var i in out) {
+				var CMDstring = 'gmetric --name="' + i + '_' + config.INSTANCE + '" --value=' + out[i] + ' --type=float --unit=n --dmax=90 --group=Chat';
+				execCMD(CMDstring, function() {
+				});
```
Comment (hakubo) [inline monitoring.js]: one tab too many?

File: mw_client.js [hunk 9]
```diff
@@ -0,0 +1,311 @@
+var monitoring = require('./monitoring.js'),
```
Comment (hakubo) [inline mw_client.js]: one var per line?

File: mw_client.js [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+var urlencode = require('./lib/mw_urlencode').urlencode;
+
+var sendRequestToMediawiki = function(method, roomId, postdata, query, handshake, callback, errorcallback) {
+	if (!errorcallback) {
+		errorcallback = function() {
+		};
+	}
+
+	if (typeof postdata == "object" && method == 'POST') {
+		logger.debug('[mw::request]', postdata);
```
Comment (hakubo) [inline mw_client.js]: two logs with the same message - would be more useful to be able to tell which is which

File: mw_client.js [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+		postdata = qs.stringify(postdata);
+		logger.debug('[mw::request]', postdata);
+	} else {
+		postdata = "";
+	}
+
+	storage.getRoomSetting(roomId, 'wgServer', function(server) {
+			logger.debug('[mw::request]', 'got server url: ', server);
```
Comment (hakubo) [inline mw_client.js]: one tab too many?

File: mw_client.js [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+						logger.error('[mw::request::updateMWAddress]', 'Old wiki address found: ' + server + ', updating to ' + redirectInfo.newServer);
+						storage.setRoomSetting(roomId, 'wgServer', redirectInfo.newServer);
+					}
+				}
+
+				/**
+				 * Make a request to MW host entrypoint
+				 */
+				function makeRequest(host) {
+					var requestUrl = 'http://' + host + '/index.php' + query + "&cb=" + Math.floor(Math.random() * 99999), // varnish appears to be caching this (at least on dev boxes) when we don't want it to... so cachebust it.;
```
Comment (hakubo) [inline mw_client.js]: url.format?

File: mw_client.js [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+		token: config.TOKEN
+	});
+
+	var userToSend = [];
+
+	for (var userName in users) {
+		userToSend.push(userName);
+	}
+
+	sendRequestToMediawiki('POST', roomId, {users: userToSend}, requestUrl, null, function(data) {
```
Comment (hakubo) [inline mw_client.js]: empty callback?

File: package.json [hunk 14]
```diff
@@ -11,13 +11,15 @@
...
-    "lodash": "^4.6.1",
+    "lodash": "4.6.1",
     "qs": "1.2.2",
     "redis": "0.12.1",
     "request": "2.70.0",
     "socket.io": "1.4.5",
     "socket.io-client": "1.4.5",
-    "underscore": "1.8.3"
+    "underscore": "1.8.3",
+    "winston": "^2.2.0",
```
Comment (hakubo) [inline package.json]: can you lock versions for these as well?

File: server.js [hunk 15]
```diff
@@ -1,46 +1,52 @@
...
-    , Backbone = require('backbone')
-    , storage = require('./storage').redisFactory()
-    , models = require('./models/models')
-    , mwBridge = require('./WMBridge.js').WMBridge
-    , loggerModule = require('./logger.js')
-    , tracker = require('./tracker.js')
-    , monitoring = require('./monitoring.js')
-    , logger = loggerModule.logger
-    , keys = require("./keys_helpers");
+var app = require('express')(),
```
Comment (hakubo) [inline server.js]: one var per line?

File: server.js [hunk 16]
```diff
@@ -1,46 +1,52 @@
...
+	server = require('http')
+		.Server(app),
+	io = require('socket.io')(server),
+	jade = require('jade'),
+	_ = require('underscore')
+		._,
```
Comment (hakubo) [inline server.js]: isin't _ a default export?

File: server.js [hunk 17]
```diff
@@ -1,46 +1,52 @@
...
+	storage = require('./storage')
+		.redisFactory(),
+	models = require('./models/models'),
+	mediawikiClient = require('./mw_client.js')
+		.MediawikiClient,
```
Comment (hakubo) [inline server.js]: can this be a default export?

File: server.js [hunk 18]
```diff
@@ -87,140 +92,140 @@ storage.getRuntimeStats(function(data) {
...
 	var sessionId = false;
-	if(typeof client.myUser != 'undefined' && typeof client.myUser.get != 'undefined'){
-		sessionId = sessionIdsByKey[keys.getKey_userInRoom(client.myUser.get('name'), client.roomId)];
+	if (typeof socket.myUser != 'undefined' && typeof socket.myUser.get != 'undefined') {
+		sessionId = getSessionIdFor(socket.roomId, socket.myUser.get('name'));
 	}
-	if(sessionId === false || (typeof sessionId == "undefined") || (sessionId != client.sessionId)){
-		client.json.send({
-				event: 'forceReconnect'
+	if (sessionId === false || (typeof sessionId == "undefined") || (sessionId != socket.sessionId)) {
```
Comment (hakubo) [inline server.js]: single quotes?

File: server.js [hunk 19]
```diff
@@ -87,140 +92,140 @@ storage.getRuntimeStats(function(data) {
...
 		// Message ignored. Log the reason.
-		if(sessionId === false || typeof sessionId == "undefined"){
-				logger.debug('[messageDispatcher]', "GOT A MESSAGE WITH NO socket.io sessionId. ASSUMING THEY ARE BANNED AND SKIPPING.");
-		} else if(sessionId != client.sessionId){
-				var msg = "Got a message from a user with a mismatched client-id. This implies that the user sending the message ";
-				msg += "(" + client.myUser.get('name');
-				msg += ") has connected from a newer browser and the old browser which sent the message has been closed ";
-				msg += "and is in the process of having its connection closed.";
-				logger.debug('[messageDispatcher]', msg);
+		if (sessionId === false || typeof sessionId == "undefined") {
```
Comment (hakubo) [inline server.js]: single quotes?

File: server.js [hunk 20]
```diff
@@ -87,140 +92,140 @@ storage.getRuntimeStats(function(data) {
...
+			socket.logger.debug('[handleSocketMessage]', "GOT A MESSAGE WITH NO socket.io sessionId. ASSUMING THEY ARE BANNED AND SKIPPING.");
+		} else if (sessionId != socket.sessionId) {
+			var msg = "Got a message from a user with a mismatched client-id. This implies that the user sending the message ";
```
Comment (hakubo) [inline server.js]: quotes?

File: server.js [hunk 21]
```diff
@@ -320,137 +328,123 @@ function authConnection(socket, next){
 				}
 			}, errback);
 		} else {
-			logger.error('[authConnection]', "User failed authentication. Error from server was: " + data.errorMsg);
-//			sendInlineAlertToClient(client, data.error, data.errorWfMsg, data.errorMsgParams);
-			logger.debug('[authConnection]', "Entire data was: ");
-			logger.debug('[authConnection]', data);
-			next(new Error('User failed authentication (1)'));
+			socket.logger.error('[authConnection]', "Authentication failed. Reason: " + data.errorMsg);
+//			sendInlineAlertToClient(socket, data.error, data.errorWfMsg, data.errorMsgParams);
```
Comment (hakubo) [inline server.js]: commented code - remove?

File: server_api.js [hunk 22]
```diff
@@ -8,43 +8,45 @@
  * This API server is used for filling requests from the MediaWiki server.
  */
 
-var config = require("./server_config.js"); // our node-chat config
-var http = require("http");
-var urlUtil = require('url');
-var storage = require('./storage').redisFactory();
-var logger = require('./logger').logger;
+var config = require("./server_config.js"),
```
Comment (hakubo) [inline server_api.js]: one var?

File: server_api.js [hunk 23]
```diff
@@ -8,43 +8,45 @@
...
  * resulting callbacks.
  */
-function apiDispatcher(req, res, reqData, successCallback, errorCallback){
-	if(reqData.query.func){
+function apiDispatcher(req, res, reqData, successCallback, errorCallback) {
+	if (reqData.query.func) {
 		var func = reqData.query.func.toLowerCase();
 
-		if(func == "getdefaultroomid"){
+		if (func == "getdefaultroomid") {
```
Comment (hakubo) [inline server_api.js]: quotes?

File: server_config.js [hunk 24]
```diff
@@ -1,31 +1,34 @@
-var md5 = require("./lib/md5.js").md5;
-var _ = require('lodash');
+var md5 = require("./lib/md5.js")
+		.md5,
```
Comment (hakubo) [inline server_config.js]: this can be default export

File: storage.js [hunk 25]
```diff
@@ -1,9 +1,9 @@
-var logger = require('./logger').logger;
-var keys = require('./keys_helpers')
-var config = require("./server_config.js"); // our node-chat config
-var _ = require('lodash')
+var logger = require('./logger').forModule(module);
+var keys = require('./keys_helpers'),
+	config = require("./server_config.js"),
```
Comment (hakubo) [inline storage.js]: one var

File: storage.js [hunk 26]
```diff
@@ -180,144 +181,145 @@ RedisStorage.prototype = {
...
-				self.getRoomData(roomId, 'wgCityId', function(wgCityId) {
-					// Only purge the users for the current instance.
-					// NOTE: This functionality is designed to mirror what's in ChatHelper.php::getServer().
-					var roomInstance = ((wgCityId % config.INSTANCE_COUNT) + 1);
-					if(roomInstance == instanceNumber){
-						logger.debug("\tCleaning users out of room with key: " + usersInRoomKey);
-						self._del(usersInRoomKey, self._redis.print, null, self._redis.print);
-					}
+		self._keys(keys.getKeyPrefix_usersInRoom() + ":*", function(data) {
+				_.each(data, function(usersInRoomKey) {
```
Comment (hakubo) [inline storage.js]: ERR_TOO_MANY_TABS

File: tracker.js [hunk 27]
```diff
@@ -1,51 +1,52 @@
...
+	request = require('request');
+var logger = require('./logger').forModule(module);
 
-var config = require("./server_config.js");
-var qs = require('qs');
-var request = require('request');
-var logger = require('./logger.js').logger;
-var trackId = 0;
-var lastTs = 0;
+var trackId = 0,
```
Comment (hakubo) [inline tracker.js]: one var?

File: mw_client.js (hunk 11 above)
Comment (pchojnacki) [inline mw_client.js]: I havent done this as this funciton is very long

File: mw_client.js (hunk 10 above)
Comment (pchojnacki) [inline mw_client.js]: I've added additional string suffix to the other logger.debug

File: .editorconfig [hunk 28]
```diff
@@ -0,0 +1,10 @@
+root= true
```
Comment (hakubo) [inline .editorconfig]: missing space :)

File: keys_helpers.js [hunk 29]
```diff
@@ -1,5 +1,5 @@
-var config = require('./server_config');
-var md5 = require("./lib/md5.js").md5;
+var config = require('./server_config'),
```
Comment (hakubo) [inline keys_helpers.js]: one var?

File: lib/md5.js [hunk 30]
```diff
@@ -1,9 +1,9 @@
...
 
-var md5 = function (plainText) {
+var md5 = function(plainText) {
 	var md5Hash = crypto.createHash('md5');
 	md5Hash.update(plainText);
 	return md5Hash.digest('hex');
-}
+};
 
 module.exports.md5 = md5;
```
Comment (hakubo) [inline lib/md5.js]: export default?

File: monitoring.js [hunk 31]
```diff
@@ -1,93 +1,96 @@
...
 			out.runtime = (parseInt(new Date().getTime()) - parseInt(runtime_data.laststart) + parseInt(runtime_data.runtime) );
-			out.avg_runtime = out.runtime/(parseInt(runtime_data.startcount) + 1);
-			storage.getUserCount(function(usercount_data) {
+			out.avg_runtime = out.runtime / (parseInt(runtime_data.startcount) + 1);
+			storage.getTotalUserCount(function(usercount_data) {
 				out.usercount = usercount_data;
 				execCMD("ps -opcpu -p " + process.pid + " | grep -v CPU", function(node_cpu) {
 					out.node_cpup = node_cpu;
-					execCMD( "ps -opcpu -p " + redis_pid + " | grep -v CPU", function(redis_cpu) {
+					execCMD("ps -opcpu -p " + redis_pid + " | grep -v CPU", function(redis_cpu) {
```
Comment (hakubo) [inline monitoring.js]: please grep for "

File: monitoring.js [hunk 32]
```diff
@@ -1,93 +1,96 @@
...
+	var stats = [];
+	storage.info(function(storagedata) {
+		var data = storagedata.split("\r\n");
+		for (var i = 0; i < data.length; i++) {
+			var temp = data[i].split(":");
+			if (temp.length == 2) {
+				stats[temp[0]] = temp[1];
+			}
+		}
+		// Holy log-spam batman! :)
```
Comment (hakubo) [inline monitoring.js]: can we remove that?

File: mw_client.js (hunk 13 above)
Comment (pchojnacki) [inline mw_client.js]: I've logged the callback results

**Comment by pchojnacki** [comment]
ping @hakubo

**Comment by pchojnacki** [comment]
LGTM

**Comment by pchojnacki** [comment]
@ludwikkazmierczak @hakubo and me did test it and deemed it worthy of merging

**Pull request merged**